### ISSUE TITLE
Mypy allows Any to be passed into a typed function

### ISSUE THREAD

**Issue by yalwan-sage** [python/mypy#15550]
**Bug Report**

Mypy allows `Any` to be passed into a typed function

**To Reproduce**

```python
import typing

def sample(a: str) -> None:
    pass

badly_typed: typing.Any = 1

sample(badly_typed)
```

**Expected Behavior**

`Argument "a" to "sample" has incompatible type "Any"; expected "str"  [arg-type]`

**Actual Behavior**

```bash
$ mypy demo.py 
Success: no issues found in 1 source file
$
```

**Your Environment**
```bash
$ pip list installed | grep mypy 
mypy              1.4.1
mypy-extensions   1.0.0
$ python --version
Python 3.11.2
$ 
```

**Comment by AlexWaygood** [comment]
`--disallow-any-expr` would do it. If the object is inferred as `Any` due to mypy not being able to follow the import, `--disallow-any-unimported` would also do it.

**Comment by yalwan-sage** [comment]
> `--disallow-any-expr` would do it

This is rather broad, and not what I'd want for the example given. It seems like what I'd actually be interested in, is a flag which disallows this

> `Any` is a subtype of all other types

Is there such a flag?

**Comment by AlexWaygood** [comment]
> It seems like what I'd actually be interested in, is a flag which disallows this
> 
> > `Any` is a subtype of all other types
> 
> Is there such a flag?

If that's what you want, just use `object` instead of `Any` for your type annotation. Or if the object comes from somewhere else, use `typing.cast` to change the inferred type to `object`

**Comment by yalwan-sage** [comment]
The problem is that you have to know that a particular operation will give an inferred type of `Any` apriori in order to cast it so that it will then be caught by the type checker (which is what I actually want here, for such inputs to get flagged by the type checker)

**Comment by AlexWaygood** [comment]
MyPy's behaviour here is correct according to the spec, I'm afraid, and I doubt we'll be adding another configuration flag so that we can provide the behaviour you'd like here
